proposal.md

```diff
@@ -0,0 +1,59 @@
+## Flexible PipelineAgent IO
+
+Rename ``VoicePipelineAgent`` to ``PipelineAgent``
+
+STT -> LLM -> TTS
+
+```python
+# rtc.AudioStream already satisfies this interface
+class AudioInput(Protocol):
+    async def __anext__(self) -> rtc.AudioFrame | rtc.AudioFrameEvent: ...
+
+
+# rtc.AudioSource already satisfies this interface
+class AudioOutput(Protocol):
+    async def capture_frame(self, frame: rtc.AudioFrame) -> None: ...
+
+    async def wait_for_playout(self) -> None: ...
+
+    def clear_queue(self) -> None: ...
+
+
+class PipelineIO(ABC):
+
+    def before_stt_node(self, source: AsyncIterator[rtc.AudioFrame]) -> AsyncIterator[rtc.AudioFrame]:
+        return source
+
+    def after_stt_node(self, source: AsyncIterator[SpeechEvent]) -> AsyncIterator[SpeechEvent]:
+        return source
+
+    def before_llm_node(self, chat_ctx: ChatContext) -> AsyncIterator[ChatChunk] | None:
+        return None
+
+    def after_llm_node(self, source: AsyncIterator[ChatChunk]) -> AsyncIterator[ChatChunk]:
+        return source
+
+    def before_tts_node(self, source: AsyncIterator[str]) -> AsyncIterator[rtc.AudioFrame] | None:
+        return source
+
+    def after_tts_node(self, source: AsyncIterator[rtc.AudioFrame]) -> AsyncIterator[rtc.AudioFrame]:
+        return source
```

Comment:
Not my wheelhouse, but is this in any way more convenient or idiomatic than if each pipeline stage managed its own pre/post transform callbacks? E.G.:

```python
def passthrough_audio(source: AsyncIterator[rtc.AudioFrame]) -> AsyncIterator[rtc.AudioFrame]:
    return source

def filter_swearwords(source: AsyncIterator[SpeechEvent]) -> AsyncIterator[SpeechEvent]:
    return source

agent = PipelineAgent(
    stt=STT(pre=passthrough_audio, post=filter_swearwords)
    ...
)
```

Reply:
The scope is different, ideally, we can add new parameters like speech_id inside each step like before_tts_node, ...